File: clients/typescript/src/entity-query.ts
import { create, fromBinary, toBinary } from "@bufbuild/protobuf";
import type { DescMessage, MessageShape } from "@bufbuild/protobuf";
import { Access, QueryDescriptorSchema } from "@ecs/protos/ecs/protocol/v1/query_pb.js";
import type { QueryDescriptor } from "@ecs/protos/ecs/protocol/v1/query_pb.js";
import type { ComponentBatch, SystemInvocation } from "@ecs/protos/ecs/protocol/v1/tick_pb.js";
import type { ComponentTypeDeclaration } from "@ecs/protos/ecs/protocol/v1/schema_pb.js";
import { componentType } from "./component-type.js";
import { encodeBatch, type ComponentColumn } from "./protocol/index.js";
import { Query, SchemaBindings, type ComponentAccess } from "./query-types.js";

/**
 * A typed view over the entities a system was leased for this tick.
 *
 * Declared in the system constructor, bound to dense type ids during the registration
 * handshake, then repopulated each tick from the invocation's component batches.
 *
 * A system cannot look up an arbitrary entity: it sees exactly the slice its lease
 * covers. To dereference a reference component, declare a second query matching the
 * target entities and index it — both are filled from the same invocation, so that
 * costs bandwidth rather than a round trip.
 */
export class EntityQuery {
  private readonly required: ComponentAccess[] = [];
  private readonly optional: ComponentAccess[] = [];
  private readonly excluded: ComponentAccess[] = [];
  private readonly tags: ComponentAccess[] = [];
  private frozen = false;

  private bindings = new SchemaBindings();
  private requiredIds: number[] = [];
  private optionalIds: number[] = [];
  private excludedIds: number[] = [];
  private tagIds: number[] = [];
  private declaredTypes = new Set<number>();
  private writableTypes = new Set<number>();

  private data = new Map<number, Map<bigint, Uint8Array>>();
  private resolvedTags = new Map<number, number[]>();
  private matched: bigint[] = [];
  private mutations = new Map<number, Map<bigint, Uint8Array>>();

  // ── Declaration ─────────────────────────────────────────────

  /** Adds a required component. An entity must have it to match. */
  with(access: ComponentAccess): this {
    this.throwIfFrozen();
    this.required.push(access);
    return this;
  }

  /** Adds optional components. An entity must have at least one to match. */
  withAny(...accesses: ComponentAccess[]): this {
    this.throwIfFrozen();
    this.optional.push(...accesses);
    return this;
  }

  /** Excludes entities carrying the component. */
  without(schema: DescMessage): this {
    this.throwIfFrozen();
    this.excluded.push(Query.read(schema));
    return this;
  }

  /**
   * Joins through the type system: an entity matches when it carries at least one
   * component whose *type entity* has this tag. The matching types are resolved by the
   * coordinator every tick, so a type registered later is picked up without this query
   * changing. Tag joins are read-only.
   */
  withAnyTagged(schema: DescMessage): this {
    this.throwIfFrozen();
    this.tags.push(Query.read(schema));
    return this;
  }

  // ── Binding ─────────────────────────────────────────────────

  /** Every schema this query needs the coordinator to know about. */
  declarations(): ComponentTypeDeclaration[] {
    return [...this.required, ...this.optional, ...this.excluded, ...this.tags].map(
      (a) => a.type.declaration,
    );
  }

  freeze(): void {
    this.frozen = true;
  }

  bind(bindings: SchemaBindings): void {
    this.bindings = bindings;

    this.requiredIds = this.required.map((a) => bindings.require(a.type.name));
    this.optionalIds = this.optional.map((a) => bindings.require(a.type.name));
    this.excludedIds = this.excluded.map((a) => bindings.require(a.type.name));
    this.tagIds = this.tags.map((a) => bindings.require(a.type.name));

    this.declaredTypes = new Set([...this.requiredIds, ...this.optionalIds]);
    this.writableTypes = new Set(
      [...this.required, ...this.optional]
        .filter((a) => a.access === Access.WRITE)
        .map((a) => bindings.require(a.type.name)),
    );
  }

  toDescriptor(): QueryDescriptor {
    return create(QueryDescriptorSchema, {
      required: this.required.map((a, i) => ({
        typeId: this.requiredIds[i]!,
        access: a.access,
      })),
      optional: this.optional.map((a, i) => ({
        typeId: this.optionalIds[i]!,
        access: a.access,
      })),
      excluded: this.excludedIds,
      tagged: this.tagIds.map((id) => ({ tagTypeId: id })),
    });
  }

  readNames(): string[] {
    return [...this.required, ...this.optional]
      .filter((a) => a.access !== Access.WRITE)
      .map((a) => a.type.name);
  }

  writeNames(): string[] {
    return [...this.required, ...this.optional]
      .filter((a) => a.access === Access.WRITE)
      .map((a) => a.type.name);
  }

  // ── Per-tick population ─────────────────────────────────────

  populate(columns: Map<number, ComponentColumn>, invocation: SystemInvocation): void {
    this.data = new Map();
    this.matched = [];
    this.mutations = new Map();
    this.resolvedTags = new Map();

    const taggedTypes = new Set<number>();
    for (const tagId of this.tagIds) {
      const resolved = invocation.tags.find((t) => t.tagTypeId === tagId);
      const types = resolved?.typeIds ?? [];
      this.resolvedTags.set(tagId, [...types]);
      for (const type of types) taggedTypes.add(type);
    }

    for (const [typeId, column] of columns) {
      if (
        !this.declaredTypes.has(typeId) &&
        !this.excludedIds.includes(typeId) &&
        !taggedTypes.has(typeId)
      ) {
        continue;
      }

      const byEntity = new Map<bigint, Uint8Array>();
      const count = Math.min(column.entities.length, column.rows.length);
      for (let i = 0; i < count; i++) {
        const row = column.rows[i];
        if (row !== null && row !== undefined) byEntity.set(column.entities[i]!, row);
      }
      this.data.set(typeId, byEntity);
    }

    let candidates: Set<bigint> | undefined;

    for (const typeId of this.requiredIds) {
      const byEntity = this.data.get(typeId);
      if (byEntity === undefined) return;

      candidates =
        candidates === undefined
          ? new Set(byEntity.keys())
          : intersect(candidates, new Set(byEntity.keys()));
    }

    for (const tagId of this.tagIds) {
      const withTag = new Set<bigint>();
      for (const typeId of this.resolvedTags.get(tagId) ?? []) {
        for (const entity of this.data.get(typeId)?.keys() ?? []) withTag.add(entity);
      }
      candidates = candidates === undefined ? withTag : intersect(candidates, withTag);
    }

    if (candidates === undefined) return;

    if (this.optionalIds.length > 0) {
      for (const entity of [...candidates]) {
        const any = this.optionalIds.some((id) => this.data.get(id)?.has(entity) === true);
        if (!any) candidates.delete(entity);
      }
    }

    for (const typeId of this.excludedIds) {
      for (const entity of this.data.get(typeId)?.keys() ?? []) candidates.delete(entity);
    }

    this.matched = [...candidates];
  }

  // ── Data access ─────────────────────────────────────────────

  /** Entities matching this query for the current tick. */
  get entities(): readonly bigint[] {
    return this.matched;
  }

  get<Desc extends DescMessage>(entity: bigint, schema: Desc): MessageShape<Desc> {
    const value = this.tryGet(entity, schema);
    if (value === undefined) {
      throw new Error(`Entity ${entity} has no ${schema.typeName} in this query.`);
    }
    return value;
  }

  tryGet<Desc extends DescMessage>(
    entity: bigint,
    schema: Desc,
  ): MessageShape<Desc> | undefined {
    const typeId = this.bindings.tryGetId(schema.typeName);
    if (typeId === undefined) return undefined;

    const payload = this.data.get(typeId)?.get(entity);
    return payload === undefined ? undefined : fromBinary(schema, payload);
  }

  has(entity: bigint, schema: DescMessage): boolean {
    const typeId = this.bindings.tryGetId(schema.typeName);
    return typeId === undefined ? false : this.data.get(typeId)?.has(entity) === true;
  }

  /**
   * Buffers a component write. Throws unless the type was declared with `Query.write` —
   * the client half of the borrow check, so a mistake surfaces here rather than as a
   * rejected result a network hop later.
   */
  set<Desc extends DescMessage>(
    entity: bigint,
    schema: Desc,
    component: MessageShape<Desc>,
  ): void {
    const typeId = this.bindings.require(schema.typeName);

    if (!this.writableTypes.has(typeId)) {
      throw new Error(`Cannot write ${schema.typeName}: this query declared it read-only.`);
    }

    let byEntity = this.mutations.get(typeId);
    if (byEntity === undefined) {
      byEntity = new Map();
      this.mutations.set(typeId, byEntity);
    }
    byEntity.set(entity, toBinary(schema, component));
  }

  // ── Tag joins ───────────────────────────────────────────────

  /** The component type names carrying the tag this tick. */
  taggedTypeNames(schema: DescMessage): string[] {
    const tagId = this.bindings.tryGetId(schema.typeName);
    if (tagId === undefined) return [];

    return (this.resolvedTags.get(tagId) ?? []).map((id) => this.bindings.nameOf(id));
  }

  /**
   * The components on an entity whose type carries the tag. Payloads stay raw because
   * the concrete types are only known at runtime.
   */
  *getTagged(entity: bigint, schema: DescMessage): Generator<TaggedComponent> {
    const tagId = this.bindings.tryGetId(schema.typeName);
    if (tagId === undefined) return;

    for (const typeId of this.resolvedTags.get(tagId) ?? []) {
      const payload = this.data.get(typeId)?.get(entity);
      if (payload !== undefined) {
        yield new TaggedComponent(this.bindings.nameOf(typeId), payload);
      }
    }
  }

  // ── Flush ───────────────────────────────────────────────────

  flushWrites(): ComponentBatch[] {
    const batches: ComponentBatch[] = [];

    for (const [typeId, byEntity] of this.mutations) {
      if (byEntity.size === 0) continue;

      batches.push(
        encodeBatch({
          typeId,
          entities: [...byEntity.keys()],
          rows: [...byEntity.values()],
        }),
      );
    }

    this.mutations = new Map();
    return batches;
  }

  private throwIfFrozen(): void {
    if (this.frozen) {
      throw new Error(
        "Queries must be declared in the system constructor, before it joins a world.",
      );
    }
  }
}

/** A component reached through a tag join, whose concrete type is only known at runtime. */
export class TaggedComponent {
  constructor(
    readonly typeName: string,
    readonly payload: Uint8Array,
  ) {}

  /** Decodes the payload, or returns undefined if it is a different type. */
  as<Desc extends DescMessage>(schema: Desc): MessageShape<Desc> | undefined {
    if (this.typeName !== schema.typeName) return undefined;
    return fromBinary(schema, this.payload);
  }
}

function intersect(a: Set<bigint>, b: Set<bigint>): Set<bigint> {
  const result = new Set<bigint>();
  for (const value of a) {
    if (b.has(value)) result.add(value);
  }
  return result;
}

export { componentType };

File: clients/csharp/Client/EntityQuery.cs
using Ecs.Protocol;
using Ecs.Protocol.V1;
using Google.Protobuf;

namespace Engine.Core;

/// <summary>
/// A typed view over the entities a system was leased for this tick.
/// </summary>
/// <remarks>
/// Declared in the system constructor with <c>Query.Read&lt;T&gt;</c> and
/// <c>Query.Write&lt;T&gt;</c>, bound to dense type ids during the registration
/// handshake, then repopulated each tick from the invocation's component batches.
///
/// A system cannot look up an arbitrary entity: it sees exactly the slice its lease
/// covers. To dereference a reference component, declare a second query matching the
/// target entities and index it — both queries are filled from the same invocation in
/// the same tick, so that costs bandwidth rather than a round trip.
/// </remarks>
public sealed class EntityQuery
{
    // ── Declaration (system constructor) ────────────────────────

    private readonly List<ComponentAccess> _required = new();
    private readonly List<ComponentAccess> _optional = new();
    private readonly List<ComponentAccess> _excluded = new();
    private readonly List<ComponentAccess> _tags = new();
    private bool _frozen;

    // ── Binding (registration handshake) ────────────────────────

    private SchemaBindings _bindings = new();
    private readonly HashSet<uint> _declaredTypes = new();
    private readonly HashSet<uint> _writableTypes = new();
    private uint[] _requiredIds = [];
    private uint[] _optionalIds = [];
    private uint[] _excludedIds = [];
    private uint[] _tagIds = [];

    // ── Tick state ──────────────────────────────────────────────

    private readonly Dictionary<uint, Dictionary<ulong, byte[]>> _data = new();
    private readonly Dictionary<uint, uint[]> _resolvedTags = new();
    private readonly List<Entity> _matched = new();
    private readonly Dictionary<uint, Dictionary<ulong, byte[]>> _mutations = new();

    // ── Declaration ─────────────────────────────────────────────

    /// <summary>Adds a required component. An entity must have it to match.</summary>
    public EntityQuery With(ComponentAccess access)
    {
        ThrowIfFrozen();
        _required.Add(access);
        return this;
    }

    /// <summary>Adds optional components. An entity must have at least one to match.</summary>
    public EntityQuery WithAny(params ComponentAccess[] accesses)
    {
        ThrowIfFrozen();
        _optional.AddRange(accesses);
        return this;
    }

    /// <summary>Excludes entities carrying the component.</summary>
    public EntityQuery Without<T>() where T : IMessage<T>, new()
    {
        ThrowIfFrozen();
        _excluded.Add(Query.Read<T>());
        return this;
    }

    /// <summary>
    /// Joins through the type system: an entity matches when it carries at least one
    /// component whose <em>type entity</em> has <typeparamref name="TTag"/>. The matching
    /// types are resolved by the coordinator every tick, so a type introduced later is
    /// picked up without this query changing. Tag joins are read-only.
    /// </summary>
    public EntityQuery WithAnyTagged<TTag>() where TTag : IMessage<TTag>, new()
    {
        ThrowIfFrozen();
        _tags.Add(Query.Read<TTag>());
        return this;
    }

    // ── Binding ─────────────────────────────────────────────────

    /// <summary>Every schema this query needs the coordinator to know about.</summary>
    internal IEnumerable<ComponentTypeDeclaration> Declarations() =>
        _required.Concat(_optional).Concat(_excluded).Concat(_tags).Select(a => a.Declaration);

    internal IEnumerable<string> ReadNames() =>
        _required.Concat(_optional).Where(a => !a.IsWrite).Select(a => a.Name);

    internal IEnumerable<string> WriteNames() =>
        _required.Concat(_optional).Where(a => a.IsWrite).Select(a => a.Name);

    internal void Freeze() => _frozen = true;

    /// <summary>
    /// Resolves declared type names to the ids the coordinator assigned. Called once the
    /// registration handshake completes, before the first invocation.
    /// </summary>
    internal void Bind(SchemaBindings bindings)
    {
        _bindings = bindings;

        _requiredIds = [.. _required.Select(a => bindings.Require(a.Name))];
        _optionalIds = [.. _optional.Select(a => bindings.Require(a.Name))];
        _excludedIds = [.. _excluded.Select(a => bindings.Require(a.Name))];
        _tagIds = [.. _tags.Select(a => bindings.Require(a.Name))];

        _declaredTypes.Clear();
        _declaredTypes.UnionWith(_requiredIds);
        _declaredTypes.UnionWith(_optionalIds);

        _writableTypes.Clear();
        foreach (var access in _required.Concat(_optional).Where(a => a.IsWrite))
            _writableTypes.Add(bindings.Require(access.Name));
    }

    internal QueryDescriptor ToDescriptor()
    {
        var descriptor = new QueryDescriptor();

        for (var i = 0; i < _required.Count; i++)
            descriptor.Required.Add(new Ecs.Protocol.V1.ComponentAccess
            {
                TypeId = _requiredIds[i],
                Access = _required[i].Access,
            });

        for (var i = 0; i < _optional.Count; i++)
            descriptor.Optional.Add(new Ecs.Protocol.V1.ComponentAccess
            {
                TypeId = _optionalIds[i],
                Access = _optional[i].Access,
            });

        descriptor.Excluded.AddRange(_excludedIds);
        descriptor.Tagged.AddRange(_tagIds.Select(id => new TaggedAccess { TagTypeId = id }));
        return descriptor;
    }

    // ── Per-tick population ─────────────────────────────────────

    /// <summary>
    /// Rebuilds this query's view from the tick's invocation. The coordinator has already
    /// narrowed the slice to entities matching at least one of the system's queries; this
    /// applies each individual query's own filter.
    /// </summary>
    internal void Populate(IReadOnlyDictionary<uint, ComponentColumn> columns, SystemInvocation invocation)
    {
        _data.Clear();
        _matched.Clear();
        _mutations.Clear();
        _resolvedTags.Clear();

        var taggedTypes = new HashSet<uint>();
        foreach (var tagId in _tagIds)
        {
            var resolved = invocation.Tags.FirstOrDefault(t => t.TagTypeId == tagId);
            var types = resolved is null ? [] : resolved.TypeIds.ToArray();
            _resolvedTags[tagId] = types;
            taggedTypes.UnionWith(types);
        }

        foreach (var (typeId, column) in columns)
        {
            if (!_declaredTypes.Contains(typeId) &&
                !_excludedIds.Contains(typeId) &&
                !taggedTypes.Contains(typeId))
            {
                continue;
            }

            var byEntity = new Dictionary<ulong, byte[]>(column.Count);
            for (var i = 0; i < column.Count; i++)
            {
                if (column.Rows[i] is { } row) byEntity[column.Entities[i]] = row;
            }
            _data[typeId] = byEntity;
        }

        HashSet<ulong>? candidates = null;

        foreach (var typeId in _requiredIds)
        {
            if (!_data.TryGetValue(typeId, out var byEntity)) return;

            if (candidates is null) candidates = [.. byEntity.Keys];
            else candidates.IntersectWith(byEntity.Keys);
        }

        foreach (var tagId in _tagIds)
        {
            var withTag = new HashSet<ulong>();
            foreach (var typeId in _resolvedTags[tagId])
            {
                if (_data.TryGetValue(typeId, out var byEntity)) withTag.UnionWith(byEntity.Keys);
            }

            if (candidates is null) candidates = withTag;
            else candidates.IntersectWith(withTag);
        }

        if (candidates is null) return;

        if (_optionalIds.Length > 0)
        {
            candidates.RemoveWhere(entity =>
                !_optionalIds.Any(id => _data.TryGetValue(id, out var d) && d.ContainsKey(entity)));
        }

        foreach (var typeId in _excludedIds)
        {
            if (_data.TryGetValue(typeId, out var byEntity)) candidates.ExceptWith(byEntity.Keys);
        }

        foreach (var entity in candidates) _matched.Add(new Entity(entity));
    }

    // ── Data access ─────────────────────────────────────────────

    /// <summary>Entities matching this query for the current tick.</summary>
    public IReadOnlyList<Entity> Entities => _matched;

    public T Get<T>(Entity entity) where T : IMessage<T>, new() =>
        TryGet<T>(entity, out var component)
            ? component
            : throw new KeyNotFoundException(
                $"Entity {entity.Id} has no {ComponentType<T>.Name} in this query.");

    public bool TryGet<T>(Entity entity, out T component) where T : IMessage<T>, new()
    {
        if (_bindings.TryGetId(ComponentType<T>.Name, out var typeId) &&
            _data.TryGetValue(typeId, out var byEntity) &&
            byEntity.TryGetValue(entity.Id, out var payload))
        {
            component = new T();
            component.MergeFrom(payload);
            return true;
        }

        component = default!;
        return false;
    }

    public bool Has<T>(Entity entity) where T : IMessage<T>, new() =>
        _bindings.TryGetId(ComponentType<T>.Name, out var typeId) &&
        _data.TryGetValue(typeId, out var byEntity) &&
        byEntity.ContainsKey(entity.Id);

    /// <summary>
    /// Buffers a component write. Throws unless the type was declared with
    /// <see cref="Query.Write{T}"/> — the client half of the borrow check, so a mistake
    /// surfaces here rather than as a rejected result a network hop later.
    /// </summary>
    public void Set<T>(Entity entity, T component) where T : IMessage<T>, new()
    {
        var typeId = _bindings.Require(ComponentType<T>.Name);

        if (!_writableTypes.Contains(typeId))
            throw new InvalidOperationException(
                $"Cannot write {ComponentType<T>.Name}: this query declared it read-only.");

        if (!_mutations.TryGetValue(typeId, out var byEntity))
        {
            byEntity = new Dictionary<ulong, byte[]>();
            _mutations[typeId] = byEntity;
        }
        byEntity[entity.Id] = component.ToByteArray();
    }

    // ── Tag joins ───────────────────────────────────────────────

    /// <summary>The component types carrying <typeparamref name="TTag"/> this tick.</summary>
    public IReadOnlyList<string> TaggedTypeNames<TTag>() where TTag : IMessage<TTag>, new()
    {
        if (!_bindings.TryGetId(ComponentType<TTag>.Name, out var tagId) ||
            !_resolvedTags.TryGetValue(tagId, out var types))
        {
            return [];
        }

        return [.. types.Select(_bindings.NameOf)];
    }

    /// <summary>
    /// The components on <paramref name="entity"/> whose type carries
    /// <typeparamref name="TTag"/>. Payloads stay raw because the concrete types are only
    /// known at runtime — decode one with <see cref="TaggedComponent.As{T}"/> once its
    /// name identifies something you know.
    /// </summary>
    public IEnumerable<TaggedComponent> GetTagged<TTag>(Entity entity) where TTag : IMessage<TTag>, new()
    {
        if (!_bindings.TryGetId(ComponentType<TTag>.Name, out var tagId) ||
            !_resolvedTags.TryGetValue(tagId, out var types))
        {
            yield break;
        }

        foreach (var typeId in types)
        {
            if (_data.TryGetValue(typeId, out var byEntity) &&
                byEntity.TryGetValue(entity.Id, out var payload))
            {
                yield return new TaggedComponent(_bindings.NameOf(typeId), payload);
            }
        }
    }

    // ── Tuple iteration ─────────────────────────────────────────

    public IEnumerable<(Entity Entity, T1 C1)> Each<T1>()
        where T1 : IMessage<T1>, new()
    {
        foreach (var entity in _matched)
        {
            if (TryGet<T1>(entity, out var c1)) yield return (entity, c1);
        }
    }

    public IEnumerable<(Entity Entity, T1 C1, T2 C2)> Each<T1, T2>()
        where T1 : IMessage<T1>, new()
        where T2 : IMessage<T2>, new()
    {
        foreach (var entity in _matched)
        {
            if (TryGet<T1>(entity, out var c1) && TryGet<T2>(entity, out var c2))
                yield return (entity, c1, c2);
        }
    }

    public IEnumerable<(Entity Entity, T1 C1, T2 C2, T3 C3)> Each<T1, T2, T3>()
        where T1 : IMessage<T1>, new()
        where T2 : IMessage<T2>, new()
        where T3 : IMessage<T3>, new()
    {
        foreach (var entity in _matched)
        {
            if (TryGet<T1>(entity, out var c1) && TryGet<T2>(entity, out var c2) &&
                TryGet<T3>(entity, out var c3))
            {
                yield return (entity, c1, c2, c3);
            }
        }
    }

    public IEnumerable<(Entity Entity, T1 C1, T2 C2, T3 C3, T4 C4)> Each<T1, T2, T3, T4>()
        where T1 : IMessage<T1>, new()
        where T2 : IMessage<T2>, new()
        where T3 : IMessage<T3>, new()
        where T4 : IMessage<T4>, new()
    {
        foreach (var entity in _matched)
        {
            if (TryGet<T1>(entity, out var c1) && TryGet<T2>(entity, out var c2) &&
                TryGet<T3>(entity, out var c3) && TryGet<T4>(entity, out var c4))
            {
                yield return (entity, c1, c2, c3, c4);
            }
        }
    }

    // ── Flush ───────────────────────────────────────────────────

    /// <summary>Encodes and clears the buffered writes for this tick's result.</summary>
    internal List<ComponentBatch> FlushWrites()
    {
        var batches = new List<ComponentBatch>(_mutations.Count);

        foreach (var (typeId, byEntity) in _mutations)
        {
            if (byEntity.Count == 0) continue;

            batches.Add(ComponentBatchCodecs.Encode(new ComponentColumn(
                typeId,
                [.. byEntity.Keys],
                [.. byEntity.Values])));
        }

        _mutations.Clear();
        return batches;
    }

    private void ThrowIfFrozen()
    {
        if (_frozen)
            throw new InvalidOperationException(
                "Queries must be declared in the system constructor, before it joins a world.");
    }
}

/// <summary>
/// A component reached through a tag join, whose concrete type is only known at runtime.
/// </summary>
public readonly record struct TaggedComponent(string TypeName, byte[] Payload)
{
    /// <summary>Decodes the payload as <typeparamref name="T"/>, or null if it is a different type.</summary>
    public T? As<T>() where T : class, IMessage<T>, new()
    {
        if (TypeName != ComponentType<T>.Name) return null;

        var component = new T();
        component.MergeFrom(Payload);
        return component;
    }
}
